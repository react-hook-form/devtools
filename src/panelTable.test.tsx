import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import PanelTable from './panelTable';

describe('PanelTable', () => {
  it('should render collapsed', () => {
    const { asFragment } = render(
      <PanelTable
        isNative={false}
        errorMessage="test"
        errorType="test"
        hasError
        type="test"
        isTouched
        isDirty
        readFormStateRef={{ current: { touched: false } }}
        index={0}
        fieldsValues={{}}
        name="test"
        collapseAll={false}
        refObject={{}}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render string field values correctly', () => {
    const { getByTitle, getByTestId } = render(
      <PanelTable
        isNative={false}
        errorMessage="test"
        errorType="test"
        hasError
        type="test"
        isTouched
        isDirty
        readFormStateRef={{ current: { touched: false } }}
        index={0}
        fieldsValues={{ test: 'RHF' }}
        name="test"
        collapseAll={false}
        refObject={{}}
      />,
    );

    fireEvent.click(getByTitle('Toggle field table'));

    expect(getByTestId('test-field-value').textContent).toBe('RHF');
  });

  it('should render number field values correctly', () => {
    const { getByTitle, getByTestId } = render(
      <PanelTable
        isNative={false}
        errorMessage="test"
        errorType="test"
        hasError
        type="test"
        isTouched
        isDirty
        readFormStateRef={{ current: { touched: false } }}
        index={0}
        fieldsValues={{ test: 1234 }}
        name="test"
        collapseAll={false}
        refObject={{}}
      />,
    );

    fireEvent.click(getByTitle('Toggle field table'));

    expect(getByTestId('test-field-value').textContent).toBe('1234');
  });

  it('should render boolean field values correctly', () => {
    const { getByTitle, getByTestId } = render(
      <PanelTable
        isNative={false}
        errorMessage="test"
        errorType="test"
        hasError
        type="test"
        isTouched
        isDirty
        readFormStateRef={{ current: { touched: false } }}
        index={0}
        fieldsValues={{ test: false }}
        name="test"
        collapseAll={false}
        refObject={{}}
      />,
    );

    fireEvent.click(getByTitle('Toggle field table'));

    expect(getByTestId('test-field-value').textContent).toBe('false');
  });

  it('should render null field values correctly', () => {
    const { getByTitle, getByTestId } = render(
      <PanelTable
        isNative={false}
        errorMessage="test"
        errorType="test"
        hasError
        type="test"
        isTouched
        isDirty
        readFormStateRef={{ current: { touched: false } }}
        index={0}
        fieldsValues={{ test: null }}
        name="test"
        collapseAll={false}
        refObject={{}}
      />,
    );

    fireEvent.click(getByTitle('Toggle field table'));

    expect(getByTestId('test-field-value').textContent).toBe('null');
  });

  it('should render object field values correctly', () => {
    const { getByTitle, getByTestId } = render(
      <PanelTable
        isNative={false}
        errorMessage="test"
        errorType="test"
        hasError
        type="test"
        isTouched
        isDirty
        readFormStateRef={{ current: { touched: false } }}
        index={0}
        fieldsValues={{ test: { deeply: { nested: 'value' } } }}
        name="test"
        collapseAll={false}
        refObject={{}}
      />,
    );

    fireEvent.click(getByTitle('Toggle field table'));

    expect(getByTestId('test-field-value').textContent).toMatchInlineSnapshot(`
      "{
        \\"deeply\\": {
          \\"nested\\": \\"value\\"
        }
      }"
    `);
  });

  it('should render non-serializable field values correctly', () => {
    const circular: any = {};
    circular.circular = circular;

    const { getByTitle, getByTestId } = render(
      <PanelTable
        isNative={false}
        errorMessage="test"
        errorType="test"
        hasError
        type="test"
        isTouched
        isDirty
        readFormStateRef={{ current: { touched: false } }}
        index={0}
        fieldsValues={{ test: circular }}
        name="test"
        collapseAll={false}
        refObject={{}}
      />,
    );

    fireEvent.click(getByTitle('Toggle field table'));

    expect(getByTestId('test-field-value').textContent).toMatchInlineSnapshot(
      `"[Nested Object]"`,
    );
  });

  it('should not render undefined field values', () => {
    const { getByTitle, queryByTestId } = render(
      <PanelTable
        isNative={false}
        errorMessage="test"
        errorType="test"
        hasError
        type="test"
        isTouched
        isDirty
        readFormStateRef={{ current: { touched: false } }}
        index={0}
        fieldsValues={{ test: undefined }}
        name="test"
        collapseAll={false}
        refObject={{}}
      />,
    );

    fireEvent.click(getByTitle('Toggle field table'));

    expect(queryByTestId('test-field-value')).toBeNull();
  });
});
