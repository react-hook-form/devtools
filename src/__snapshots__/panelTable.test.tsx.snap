// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PanelTable should render collapsed 1`] = `
<DocumentFragment>
  <div
    style="opacity: 1; transition: all 0.3s ease-in 0s;"
  >
    <table
      class="css-vv52sx"
      style="padding: 10px 10px 10px; width: 100%; transition: .3s all; border-left: 2px solid #bf1650;"
    >
      <thead>
        <tr>
          <td
            style="width: 100px; line-height: 22px;"
            valign="top"
          >
            <button
              class="css-4vv12s"
              style="border: 1px solid #516391; border-radius: 2px; padding: 3px 5px; display: inline-block; font-size: 10px; line-height: 12px; width: 20px; text-align: center; margin-right: 10px;"
              title="Toggle field table"
              type="button"
            >
              +
            </button>
            <button
              class="css-4vv12s"
              style="border: 1px solid #516391; border-radius: 2px; padding: 3px 5px; display: inline-block; font-size: 10px; line-height: 12px; text-align: center; margin-right: 10px; cursor: not-allowed; background: rgb(81, 99, 145);"
              title="Scroll into view"
              type="button"
            >
              Custom
            </button>
          </td>
          <td
            style="display: block; max-width: 100px;"
          >
            <p
              style="margin: 0px; padding: 0px; font-size: 14px; line-height: 22px;"
            >
              test
            </p>
          </td>
        </tr>
      </thead>
    </table>
  </div>
</DocumentFragment>
`;
